5단계 - 자동차 경주(리팩터링)

1stpie opened the pull request:
안녕하세요. 늦었지만 잘 부탁 드리겠습니다.
좋은 리뷰 항상 감사 드립니다. 👍

1stpie committed:
test(core): move getWinners Test function

1stpie committed:
refactor(core): move Cars, Car functions to Race class

1stpie committed:
test(core): update test cases

1stpie committed:
refactor(core): change getWinners logic

1stpie committed:
style(core): change to expression

1stpie committed:
doc(readme): update requirements

1stpie committed:
refactor(core): refactoring to mvc pattern

move classes packages
move test packages
add RandomValueGenerator and test
add RacingController

ddaaac commented on src/main/kotlin/study/racingcar/controller/RacingController.kt:
```diff
@@ -0,0 +1,32 @@
+        resultView.displayWinners(racingGame.getWinners().list.map { it.name })
+    }
+
+    private fun convertCarsToList(cars: Cars): List<Pair<String, Int>> {
+        return cars.list.map { car -> Pair(car.name, car.position) }
```
자료구조가 중첩되면 코드가 복잡하게 느껴질 수도 있을 것 같아요.
List + Pair 같은 경우에 Map으로 한 번에 표현이 가능한데요 이 방법도 고려해보시면 좋을 것 같습니다.

ddaaac commented on src/main/kotlin/study/racingcar/domain/RacingGame.kt:
```diff
@@ -0,0 +1,51 @@
+    }
+
+    fun move(car: Car, randomValue: Int): Car = if (randomValue >= MOVABLE_BASE_NUMBER) {
+        car.copy(position = car.position + 1)
```
개인적으로 도메인 클래스에 data class를 사용하지 않는 이유인데요, `cars.move()`라는 메서드가 있다고 가정하고 지금 코드와 비교해보시면 좋을 것 같아요.
나중에 자동차를 한 칸 전진해야하는 로직을 다른 곳에도 작성해야될 때가 있을텐데요, 그 때마다 copy를 사용하게 된다면 나중에 한 칸 전진하는 로직을 수정할 때 정말 힘들어집니다.
어디서 position을 변경하는 코드가 있는지 모든 코드를 보면서 찾아야하거든요.
이번 교육과정에서는 도메인 로직은 도메인 내부에 두는 습관을 길러보시면 좋을 것 같습니다!

ddaaac commented on src/main/kotlin/study/racingcar/domain/RacingGame.kt:
```diff
@@ -0,0 +1,51 @@
+        cars.list.map { move(it, randomValueGenerator()) }
+    )
+
+    private fun isWin(car1: Car, other: Car) = car1.position > other.position
+
+    fun getWinners(): Cars = Cars(
+        cars.list.filter { car ->
+            cars.list.none { other -> car != other && isWin(other, car) }
+        }
+    )
```
이런 로직들에도 동일한 코멘트를 드릴 수 있겠네요.

ddaaac reviewed (changes requested):
안녕하세요 보민님 👋 

마지막 단계까지 구현 잘 해주셨습니다.

몇 가지 코멘트 남겼으니 확인해주세요~

1stpie committed:
refactor(core): change Car, Cars to class, move functions

change Car, Cars data class to class
move Car, Cars domain functions
change run() function name to raceStart

1stpie committed:
test(core): refactor test cases

1stpie committed:
refactor(core): change List<Pair> to Map

ddaaac reviewed (approved):
안녕하세요 보민님 👋 

자동차 경주 끝까지 완주하시느라 정말 고생 많으셨습니다.

다음 미션들도 화이팅하세요 💪
